src/zmalloc.c

```diff
@@ -87,10 +87,22 @@ void zlibc_free(void *ptr) {
 #define dallocx(ptr,flags) je_dallocx(ptr,flags)
 #endif
 
-#define update_zmalloc_stat_alloc(__n) atomicIncr(used_memory,(__n))
-#define update_zmalloc_stat_free(__n) atomicDecr(used_memory,(__n))
+#define update_zmalloc_stat_alloc(__n) used_memory_tls[thread_index] += (__n)
+#define update_zmalloc_stat_free(__n) used_memory_tls[thread_index] -= (__n)
 
-static serverAtomic size_t used_memory = 0;
+/* A thread-local storage which keep the current thread's index in the
+ * used_memory_tls array. */
+static __thread int thread_index;
+/* MAX_THREADS_NUM = IO_THREADS_MAX_NUM(128) + BIO threads(3) + main thread(1). */
+#define MAX_THREADS_NUM 132
```

Comment:
```suggestion
#define MAX_THREADS_NUM (IO_THREADS_MAX_NUM + BIO_THREAD_COUNT + 1)
```

Can we do something like this? I think we need to implement a new define for BIO thread count. Then we don't have to worry about this changing.

Reply:
`IO_THREADS_MAX_NUM` is defined in `networking.c` and `BIO_THREAD_COUNT` has a similar define `BIO_WORKER_NUM` in `bio.c`. How do we reuse the define in the `zmalloc.c` ? It is a very low level API.

Reply:
Yeah there is a layering concern here. Can you add a `c_assert` to make sure the condition of `(IO_THREADS_MAX_NUM + BIO_THREAD_COUNT + 1) < IO_THREADS_MAX_NUM` is not violated?

Reply:
btw, if we go with this route, we should probably introduce a new header to host all thread local storage related definition. It feels a bit out of place to define `MAX_THREADS_NUM` in zmalloc.c. We might have other consumers of this macro in the future. I wonder if we should call the new header `system.h`, which would be at the lowest layer.